### ISSUE TITLE
cobbler update mac_address/ip_address/dns_name too slow when you have more than 500 nodes

### ISSUE THREAD

**Issue by parkour99** [cobbler/cobbler#3520]
### Steps to reproduce

if hundreds of nodes have already been provisioned by cobbler, then we expand dozens of nodes, we add each node by cobbler,
cobbler will make a rpc call the method: <modify_interface> to update interfaces data.
I found that the 'modify_interface' rpc call will take 10~20 seconds for each new node. 
if we add 100 nodes by synchronously, it will take  1000-2000s. It wastes too much time

### Expected behavior
A fast **modify_interface** rpc call

### Cobbler version

````
# cobbler version
Cobbler 2.8.4
  source: ?, ?
  build time: Thu Nov 14 17:39:10 2019
````

### Operating system
>=centos7.4

### Cobbler log
Each node has12 interfaces, so the node will have 12 mac_address

````
305501 Thu Nov 23 22:15:07 2023 - DEBUG | authorize; ['cobbler', 'modify_system', <cobbler.item_system.System instance at 0xffffacec6e18>, 'modify_inte        rface', True]
 305502 Thu Nov 23 22:15:07 2023 - DEBUG | REMOTE cobbler authorization result: True; user(?)
 305503 Thu Nov 23 22:15:07 2023 - INFO | static; ['system']
 305504 Thu Nov 23 22:15:07 2023 - INFO | static; ['system']
 305505 Thu Nov 23 22:15:07 2023 - INFO | macaddress; ['system']
 305506 Thu Nov 23 22:15:07 2023 - INFO | find_items; ['system']
 305507 Thu Nov 23 22:15:10 2023 - INFO | macaddress; ['system']
 305508 Thu Nov 23 22:15:10 2023 - INFO | find_items; ['system']
 305509 Thu Nov 23 22:15:12 2023 - INFO | netmask; ['system']
 305510 Thu Nov 23 22:15:12 2023 - INFO | macaddress; ['system']
 305511 Thu Nov 23 22:15:12 2023 - INFO | find_items; ['system']
 305512 Thu Nov 23 22:15:14 2023 - INFO | macaddress; ['system']
 305513 Thu Nov 23 22:15:14 2023 - INFO | find_items; ['system']
 305514 Thu Nov 23 22:15:16 2023 - INFO | static; ['system']
 305515 Thu Nov 23 22:15:16 2023 - INFO | macaddress; ['system']
 305516 Thu Nov 23 22:15:16 2023 - INFO | find_items; ['system']
 305517 Thu Nov 23 22:15:18 2023 - INFO | static; ['system']
 305518 Thu Nov 23 22:15:18 2023 - INFO | macaddress; ['system']
 305519 Thu Nov 23 22:15:18 2023 - INFO | find_items; ['system']
 305520 Thu Nov 23 22:15:20 2023 - INFO | static; ['system']
 305521 Thu Nov 23 22:15:20 2023 - INFO | macaddress; ['system']
 305522 Thu Nov 23 22:15:20 2023 - INFO | find_items; ['system']
 305523 Thu Nov 23 22:15:22 2023 - INFO | macaddress; ['system']
 305524 Thu Nov 23 22:15:22 2023 - INFO | find_items; ['system']
 305525 Thu Nov 23 22:15:25 2023 - INFO | macaddress; ['system']
 305526 Thu Nov 23 22:15:25 2023 - INFO | find_items; ['system']
 305527 Thu Nov 23 22:15:27 2023 - INFO | static; ['system']
 305528 Thu Nov 23 22:15:27 2023 - INFO | dnsname; ['system']
 305529 Thu Nov 23 22:15:27 2023 - INFO | find_items; ['system']
 305530 Thu Nov 23 22:15:27 2023 - INFO | static; ['system']
 305531 Thu Nov 23 22:15:27 2023 - INFO | static; ['system']
 305532 Thu Nov 23 22:15:27 2023 - INFO | static; ['system']
 305533 Thu Nov 23 22:15:27 2023 - INFO | static; ['system']
 305534 Thu Nov 23 22:15:27 2023 - INFO | macaddress; ['system']
 305535 Thu Nov 23 22:15:27 2023 - INFO | find_items; ['system']
 305536 Thu Nov 23 22:15:30 2023 - INFO | static; ['system']
 305537 Thu Nov 23 22:15:30 2023 - INFO | static; ['system']
 305538 Thu Nov 23 22:15:30 2023 - INFO | macaddress; ['system']
 305539 Thu Nov 23 22:15:30 2023 - INFO | find_items; ['system']
 305540 Thu Nov 23 22:15:32 2023 - INFO | ipaddress; ['system']
 305541 Thu Nov 23 22:15:32 2023 - INFO | find_items; ['system']
 305542 Thu Nov 23 22:15:32 2023 - INFO | macaddress; ['system']
 305543 Thu Nov 23 22:15:32 2023 - INFO | find_items; ['system']
 305544 Thu Nov 23 22:15:34 2023 - INFO | authenticate; ['cobbler', True]
 305545 Thu Nov 23 22:15:34 2023 - INFO | REMOTE save_item(system); user(cobbler); object_id(___NEW___system::bW1TaQwVF7Vy7leekJAqZdb/LqZ3Tk1M/Q==)
 305546 Thu Nov 23 22:15:34 2023 - DEBUG | authorize; ['cobbler', 'save_system', <cobbler.item_system.System instance at 0xffffacec6e18>, None, True]
 305547 Thu Nov 23 22:15:34 2023 - DEBUG | REMOTE cobbler authorization result: True; user(?)
 305548 Thu Nov 23 22:15:34 2023 - INFO | add_item(system); ['node-1000']
````

### Additional information
Then I dig into the problem, got the reason why it takes too much time。

```
cobbler/item_system.py 如下：

def modify_interface(self, hash):
        """
        Used by the WUI to modify an interface more-efficiently
        """
        for (key,value) in hash.iteritems():
            (field,interface) = key.split("-",1)
            field = field.replace("_","").replace("-","")
            if field == "macaddress"          : self.set_mac_address(value, interface)
            if field == "mtu"                 : self.set_mtu(value, interface)
            if field == "ipaddress"           : self.set_ip_address(value, interface)
            if field == "dnsname"             : self.set_dns_name(value, interface)
            if field == "static"              : self.set_static(value, interface)
            if field == "dhcptag"             : self.set_dhcp_tag(value, interface)
            if field == "netmask"             : self.set_netmask(value, interface)
            if field == "subnet"              : self.set_netmask(value, interface)

```
I search **find_items** in the cobbler.log, then I got the following method
```
./item_system.py:307:           
def set_dns_name(self,dns_name,interface):
    matched = self.config.api.find_items("system", {"dns_name" : dns_name})

./item_system.py:358:
def set_ip_address(self,address,interface):          
    matched = self.config.api.find_items("system", {"ip_address" : address})
    
./item_system.py:376:
def set_mac_address(self,address,interface):           
    matched = self.config.api.find_items("system", {"mac_address" : address})
```
They will all call
```
self.config.api.find_items
```
when call **set_mac_address**

dig in **find_items**

cobbler/api.py
```
    def find_items(self, what, criteria=None):
        self.log("find_items",[what])
        # defaults
        if criteria is None:
            criteria={}
        items=self._config.get_items(what)
        # empty criteria returns everything
        if criteria == {}:
            res=items
        else:
            res=items.find(return_list=True, no_errors=False, **criteria)
        return res
```
dig in "find" method
```
cobbler/collection.py
def find(self, name=None, return_list=False, no_errors=False, **kargs):
        """
        Return first object in the collection that maches all item='value'
        pairs passed, else return None if no objects can be found.
        When return_list is set, can also return a list.  Empty list
        would be returned instead of None in that case.


    for (name, obj) in self.listing.iteritems():
                    if obj.find_match(kargs, no_errors=no_errors):
                        matches.append(obj)
```
dig in "find_match" method
```
cobbler/item.py
 def find_match(self,kwargs,no_errors=False):
        # used by find() method in collection.py
        data = self.to_datastruct()
        for (key, value) in kwargs.iteritems():
            # Allow ~ to negate the compare
            if value is not None and value.startswith("~"):
                res=not self.find_match_single_key(data,key,value[1:],no_errors)
            else:
                res=self.find_match_single_key(data,key,value,no_errors)
            if not res:
                return False
                
        return True
```
find_match_single_key > __find_compare
cobbler/item.py
```
def __find_compare(self, from_search, from_obj):

        if isinstance(from_obj, basestring):
            # FIXME: fnmatch is only used for string to string comparisions
            # which should cover most major usage, if not, this deserves fixing
            if fnmatch.fnmatch(from_obj.lower(), from_search.lower()):
                return True
            else:
                return False    
```
finally, we got the real kller: 
```
fnmatch.fnmatch
```
the fnmatch's cache size:
```
_cache = {}
_MAXCACHE = 100
```
If the size of the matched cached result is more than 100, it will clear the cache. It will make match be slower.
A **fnmatch** performance test as follow:
``` test_fnmatch.py
import fnmatch

def test(size):
    for j in range(10):
        for i in range(size):
            fnmatch.fnmatch('test', str(i))

()[root@roller zstar]# python -m timeit  -v  -n 100 -r 3 -s 'from test_fnmatch import test' -- 'test(100)'
raw times: 0.164 0.162 0.16
100 loops, best of 3: 1.6 msec per loop

()[root@roller zstar]# python -m timeit  -v  -n 102 -r 3 -s 'from test_fnmatch import test' -- 'test(102)'
raw times: 4.47 4.23 4.18
102 loops, best of 3: 41 msec per loop
```
The test show it will fast when numbers is less than 100, otherwise it will take too much time with the count of the nodes increasing.

----
In conclusion, we should optimize the string match check when check whether **mac_address/ip_address/dns_name** is duplicate or not.
Maybe only save/update **mac_address/ip_address/dns_name** into a big cache when add or remove system is a simple better choice.

I hope the community can find a better method to reduce **mac_address/ip_address/dns_name** check time, and avoid use **fnmatch** directly in the large nodes cluster.

**Comment by SchoolGuy** [comment]
@parkour99 Thanks a lot for this detailed bug report. This is really helpful. I will try and address this as soon as time allows. Sadly atm I am busy with a serverroom move at work. I hope to get to it during my vacation or in the new year.
